FALCON-1839 Test case for APIs for entities scheduled on native scheduler

pragya-mittal opened the pull request.

deepakbarr commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,176 @@
+                }
+            }
+            LOGGER.info(type + " " + entityName + " still doesn't have expected status");
+            TimeUtil.sleepSeconds(5);
```
TimeUtil.sleepSeconds(5) ?? 
I think you intended TimeUtil.sleepSeconds(sleepTime) ??

deepakbarr commented on falcon-regression/merlin/src/test/java/org/apache/falcon/regression/nativeScheduler/NativeSchedulerAPITest.java:
```diff
@@ -0,0 +1,306 @@
+    private ColoHelper cluster1 = servers.get(0);
+    private String baseTestHDFSDir = cleanAndGetTestDir();
+    private String aggregateWorkflowDir = baseTestHDFSDir + "/aggregator";
+    private static final Logger LOGGER = Logger.getLogger(NativeScheduleTest.class);
```
Logger.getLogger(NativeScheduleTest.class) ?? Wrong class name?

pragya-mittal replied:
Will fix

deepakbarr commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,176 @@
+        throws InterruptedException, IOException, AuthenticationException,
+            URISyntaxException {
+        int sleep = INSTANCES_CREATED_TIMEOUT * 60 / 5;
+        waitTillInstancesAreCreated(cluster, entity, startTime, endTime, sleep);
```
INSTANCES_CREATED_TIMEOUT * 60 / 5 - This does not seem intuitive. Why divide by 5 ?

deepakbarr commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,176 @@
+        String params = "?start=" + startTime;
+        params += (endTime.isEmpty() ? "" : "&end=" + endTime);
+
+        for (int sleepCount = 0; sleepCount < totalMinutesToWait; sleepCount++) {
```
I assume the loop keeps checking the status for "totalMinutesToWait" minutes. In the end, TimeUtil.sleepSeconds(5) should ideally be TimeUtil.sleepSeconds(60) ?

pragya-mittal replied:
@deepakbarr  I have made it to 60 seconds. But IMO waiting for a minute in every cycle for instance to get materialize is not very intuitive. Here is the thing, in oozie there is materialization service running which wakes up every 300 secs(default value) and materializes instances. But this is not the case in scheduler. In native scheduler if instanceTime < now, instance should get created. I thought of putting the basic function and depending on future observation and tests I will modify.If you have any other approach , do share!

pragya-mittal replied:
murali.msse@gmail.com Thoughts please?

muraliramasami replied:
@pragya-mittal  IMO,60secs looks long. Because it may slow down our test execution. We can try with 5-10secs. Based on the future observation we can increase if it is required.

pragya-mittal replied:
@muraliramasami Changing it to 10 secs.

pisaychuk commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,176 @@
+        for (int i = 0; i < maxTries; i++) {
+            InstancesResult statusResult = cluster.getProcessHelper().getProcessInstanceStatus(entityName, params);
+            if (statusResult.getInstances() != null) {
+                if (statusResult.getInstances()[0].getStatus().name().equals(expectedStatus.toString())) {
```
we just can compare enums themselves using ==

pragya-mittal replied:
Done.

pisaychuk commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,176 @@
+        String endTime, int totalMinutesToWait)
+        throws URISyntaxException, AuthenticationException, InterruptedException, IOException {
+        String entityName = entity.getName();
+        EntityType type = entity.getEntityType();
```
we use entity type once, so we can omit the creation of variable

pisaychuk commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,176 @@
+        throws InterruptedException, IOException, AuthenticationException,
+            URISyntaxException {
+        int sleep = INSTANCES_CREATED_TIMEOUT;
+        waitTillInstancesAreCreated(cluster, entity, startTime, endTime, sleep);
```
can't we pass INSTANCES_CREATED_TIMEOUT directly ?

pragya-mittal replied:
Yes we can. Passing it directly now.

pisaychuk commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,176 @@
+import java.net.URISyntaxException;
+
+/**
+ * util functions related to instanceTest.
```
please clarify what exact test it relates to. Or describe general purpose of methods (it might be reused by another tests in future).

pragya-mittal replied:
Will add more docs.

pisaychuk commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,176 @@
+        throws URISyntaxException, AuthenticationException, InterruptedException, IOException {
+        String entityName = entity.getName();
+        EntityType type = entity.getEntityType();
+
```
We can omit 'type' variable as it is used only once.

pisaychuk commented on falcon-regression/merlin/src/test/java/org/apache/falcon/regression/nativeScheduler/NativeSchedulerAPITest.java:
```diff
@@ -0,0 +1,306 @@
+ * Schedule process via native scheduler and test available APIs.
+ */
+
+@Test(groups = {"distributed", "embedded" })
```
we can omit distributed as we consider embedded tests runnable in distributed setup also (while distributed runnable only in distributed setup)

pragya-mittal replied:
@pisaychuk @muraliramasami As going by our offline discussion, I am keeping both groups so as to ease nightly regression.

pisaychuk commented on falcon-regression/merlin/src/test/java/org/apache/falcon/regression/nativeScheduler/NativeSchedulerAPITest.java:
```diff
@@ -0,0 +1,306 @@
+
+    /*Test for instance dependency, list, params, logs, running and triage API and check that it is successful.
+     */
+    @Test(enabled = true)
```
let's clean up (enabled = true) as it is default behaviour

pragya-mittal replied:
Done.

pisaychuk commented on falcon-regression/merlin/src/test/java/org/apache/falcon/regression/nativeScheduler/NativeSchedulerAPITest.java:
```diff
@@ -0,0 +1,306 @@
+
+    /*Test for entity definition, list, dependency, update and summary API and check that it is successful.
+     */
+    @Test(enabled = true)
```
let's clean up (enabled = true) as it is default behaviour

pragya-mittal replied:
Done.

pisaychuk commented:
@pragya-mittal I have left minor comments. Everything else looks good to me.

pragya-mittal commented:
@muraliramasami  Addressed review comments. Please check.

muraliramasami commented:
@pragya-mittal +1 Patch looks good to me.

deepakbarr commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,181 @@
+        EntityType type = entity.getEntityType();
+        String params = "?start=" + startTime;
+        params += (endTime.isEmpty() ? "" : "&end=" + endTime);
+        totalMinutesToWait = (totalMinutesToWait*60)/10;
```
Not sure about the reason for
 totalMinutesToWait = (totalMinutesToWait*60)/10

pragya-mittal replied:
Currently we are passing 10 minutes which will wait for 10seconds. i.e 100 seconds < 2minutes. But if we pass 10*60/10 = 60 * 100 i.e 600 seconds = 10 minutes. So for every 10 seconds we will check for completion up to 10minutes.

deepakbarr replied:
I think I get what you are trying to do. But this does not seem intuitive. Since you are overwriting totalMinutesToWait variable (totalMinutesToWait = (totalMinutesToWait*60)/10) and now, the name totalMinutesToWait does not actually represent the total minutes wait will happen. How about something like this - 

totalSecondsToWait = totalMinutesToWait * 60;
for (int sleepSeconds = 0; sleepSeconds < totalSecondsToWait; sleepSeconds = sleepSeconds+10) {
// Check instance
 TimeUtil.sleepSeconds(10);
}

deepakbarr replied:
OR  you can use the same logic you used in lines 132-145 in waitTillInstanceReachState() in the same class (using maxTries variable).
I just want to make sure variable names are intuitive.

pragya-mittal commented:
@muraliramasami  @deepakbarr  Modified NativeInstanceUtil.waitTillInstancesAreCreated based on review comments.

pavankumar526 commented on falcon-regression/merlin/src/test/java/org/apache/falcon/regression/nativeScheduler/NativeSchedulerAPITest.java:
```diff
@@ -0,0 +1,306 @@
+    /*Test for entity definition, list, dependency, update and summary API and check that it is successful.
+     */
+    @Test
+    public void entityDefinitionListDependencyUpdateSummaryTest() throws Exception {
```
Can we separate out update and other definition checks

pavankumar526 replied:
After update we should also check whether new staging path got created or not we have to check.

pragya-mittal replied:
These are basic sanity test cases to see whether APIs are working fine with time based scheduling. I intend to add all the specific cases once data based native scheduling comes in picture. Does it make sense ?

pavankumar526 replied:
Its good to add test for delete api also

pragya-mittal replied:
Ideally removeTestClassEntities() deletes all the entities and throws exception if delete fails. That is why didnt add the test for the same. I don't see the need of explicitly adding delete test case.

pavankumar526 replied:
Ok agreed. Please do all  checks when databased support comes

pavankumar526 commented:
+1 looks good

pragya-mittal commented:
@muraliramasami @deepakbarr @pavankumar526  I have addressed all the review comments. Modified  NativeInstanceUtil.waitTillInstancesAreCreated logic (as deepak suggested) and added documentation for the same. Please check once.

deepakbarr commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,186 @@
+     * @param startTime   start time of instance
+     * @param endTime     end time of instance
+     */
+    public static void waitTillInstancesAreCreated(ColoHelper cluster, Entity entity, String startTime, String endTime)
```
When is this method called ?? I could not find any usage.

pragya-mittal replied:
I intended to use it but then didnt use it. Anyhow with data based scheduling coming into picture it will be used heavily since we need to check instance availability before going further.

deepakbarr commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,186 @@
+     * @param endTime     end time of instance
+     * @param totalMinutesToWait     total time(in minutes) to wait for instance creation
+     */
+    public static void waitTillInstancesAreCreated(ColoHelper cluster, Entity entity, String startTime,
```
Is there an assertion missing in this method. Whether instances are created are not, the method behaviour is same. Am I missing something ?

pragya-mittal replied:
Added assertion.

deepakbarr commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,186 @@
+        int totalSleepTime = totalMinutesToWait * 60;
+        int sleepTime = totalSleepTime / maxTries;
+        LOGGER.info(String.format("Sleep for %d seconds", sleepTime));
+        for (int i = 0; i < maxTries; i++) {
```
Here, we have fixed the number of tries to 50 and sleep time because consecutive checks is computed. So, If totalMinutesToWait is 10, then, instance checks with happen every 12 seconds for 50 times. If totalMinutesToWait is 1, then, instance check will happen every 1.2 seconds for 50 times.

IMO, We should have sleepTime as fixed ( say 10 seconds ) and maxTries should be computed according to totalMinutesToWait. If totalMinutesToWait is 1 minute, We will check instance every 10 seconds
 for 6 times total. Similarly, If totalMinutesToWait is 5 minute, We will again check instance every 10 seconds for 30 times total.  What do you think ?

pragya-mittal replied:
I get it you want to implement it the same way as waitTillInstancesAreCreated(). But this approach is also fine, and I don't see any specific advantage of changing the logic here. @muraliramasami Thoughts please ?

pragya-mittal replied:
The same logic is used in InstanceUtil.waitTillInstanceReachState() and it serves the required purpose.

deepakbarr replied:
In InstanceUtil.waitTillInstanceReachState(), both sleep time and retries are fixed. 
I agree that the code will work as it and serves the purpose. It's more about the reusability of waitTillInstanceReachState() method. If someone wants to use this method  and sends  parameter totalMinutesToWait = 1, The instance check will happen every 1.2 seconds which is probably not desired/intended.

muraliramasami replied:
@pragya-mittal Looks like current function serves the purpose but not generic.That is the only concern. I think we can modify the code like:
int totalSleepTime = totalMinutesToWait * 60;
int sleepTime = 10;
int maxTries = totalSleepTime / sleepTime;

@deepakbarr Can you let us know is the above approach is good? Can you share your thoughts?

pragya-mittal replied:
Making it generic now :)

pragya-mittal commented:
@muraliramasami @deepakbarr Please review. I want to commit this patch so that I can confirm native scheduler behaviour in my nightly regression.

deepakbarr commented on falcon-regression/merlin-core/src/main/java/org/apache/falcon/regression/core/util/NativeInstanceUtil.java:
```diff
@@ -0,0 +1,184 @@
+
+        int totalSleepTime = totalMinutesToWait * 60;
+        int sleepTime = 10;
+        for (int i = 0; i < totalSleepTime; i = i+10) {
```
instead of i=i+10, make it i=i+sleepTime. Rest looks good.

pragya-mittal replied:
I was expecting it. Will make the change :)

deepakbarr commented:
One minor comment. Rest looks good.

pragya-mittal commented:
@deepakbarr Please check now.

deepakbarr commented:
+1

pragya-mittal commented:
Will merge the PR tomorrow unless there are more review comments.
